File: mobile/modules/secchat-core/ios/SecchatCoreModule.swift
import ExpoModulesCore

/// Phase-3b-Stub. Die echten UniFFI-Swift-Bindings leben unter
/// `ios/Sources/secchat_core.swift` und werden von
/// `native/secchat-core/build.sh` auf macOS erzeugt. Bis dahin wirft
/// jede Methode einen expliziten Fehler, damit JS-seitige Calls laut
/// scheitern statt still falsche Werte zu liefern.
public class SecchatCoreModule: Module {
  private func notReady() throws -> Never {
    throw NSError(
      domain: "SecchatCore", code: -1,
      userInfo: [NSLocalizedDescriptionKey:
        "secchat-core native bindings not generated yet — " +
        "run native/secchat-core/build.sh first."
      ])
  }

  public func definition() -> ModuleDefinition {
    Name("SecchatCore")

    // identity
    Function("deriveHandle") { (_: Data) throws -> String in try notReady() }
    Function("deriveIdentityPrivateKey") { (_: Data) throws -> Data in try notReady() }

    // MLS
    Function("initMls") { (_: Data) throws in try notReady() }
    AsyncFunction("generateKeyPackages") { (_: Int) throws -> [Data] in try notReady() }
    AsyncFunction("createGroup") { (_: String) throws -> [String: Any] in try notReady() }
    AsyncFunction("inviteMembers") {
      (_: [String: Any], _: [Data]) throws -> [String: Any] in try notReady()
    }
    AsyncFunction("processWelcome") { (_: Data) throws -> [String: Any] in try notReady() }
    AsyncFunction("processCommit") {
      (_: [String: Any], _: Data) throws -> [String: Any] in try notReady()
    }
    AsyncFunction("encryptMessage") {
      (_: [String: Any], _: Data) throws -> Data in try notReady()
    }
    AsyncFunction("decryptMessage") {
      (_: [String: Any], _: Data) throws -> Data in try notReady()
    }

    // Tor
    Function("torInit") { () throws in try notReady() }
    AsyncFunction("torBootstrap") { () throws in try notReady() }
    AsyncFunction("torRequest") {
      (_: String, _: String, _: Data?, _: String?) throws -> [String: Any] in try notReady()
    }
    AsyncFunction("torShutdown") { () throws in try notReady() }
  }
}

File: mobile/modules/secchat-core/android/src/main/java/com/secchat/core/SecchatCoreModule.kt
package com.secchat.core

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

/**
 * Phase-3b-Stub. Die echten UniFFI-Kotlin-Bindings leben unter
 * `android/src/main/java/uniffi/secchat_core/` und werden von
 * `native/secchat-core/build.sh` erzeugt. Solange die noch nicht
 * da sind, wirft jeder Aufruf einen klaren Fehler — so bleibt die
 * Kompilierung gruen, JS-seitige Use-Versuche schlagen aber laut
 * fehl statt still falsche Werte zu liefern.
 *
 * Phase 3c ersetzt diesen Stub durch den echten Wrapper, der
 * `uniffi.secchat_core.MlsClient`/`TorClient`/... anruft.
 */
class SecchatCoreModule : Module() {
  private fun <T> notReady(): T = throw IllegalStateException(
    "secchat-core native bindings not generated yet — " +
      "run native/secchat-core/build.sh first."
  )

  override fun definition() = ModuleDefinition {
    Name("SecchatCore")

    // identity
    Function("deriveHandle") { _: ByteArray -> notReady<String>() }
    Function("deriveIdentityPrivateKey") { _: ByteArray -> notReady<ByteArray>() }

    // MLS
    Function("initMls") { _: ByteArray -> notReady<Unit>() }
    AsyncFunction("generateKeyPackages") { _: Int -> notReady<List<ByteArray>>() }
    AsyncFunction("createGroup") { _: String -> notReady<Map<String, Any>>() }
    AsyncFunction("inviteMembers") {
      _: Map<String, Any?>, _: List<ByteArray> -> notReady<Map<String, Any>>()
    }
    AsyncFunction("processWelcome") { _: ByteArray -> notReady<Map<String, Any>>() }
    AsyncFunction("processCommit") {
      _: Map<String, Any?>, _: ByteArray -> notReady<Map<String, Any>>()
    }
    AsyncFunction("encryptMessage") {
      _: Map<String, Any?>, _: ByteArray -> notReady<ByteArray>()
    }
    AsyncFunction("decryptMessage") {
      _: Map<String, Any?>, _: ByteArray -> notReady<ByteArray>()
    }

    // Tor
    Function("torInit") { notReady<Unit>() }
    AsyncFunction("torBootstrap") { notReady<Unit>() }
    AsyncFunction("torRequest") {
      _: String, _: String, _: ByteArray?, _: String? -> notReady<Map<String, Any>>()
    }
    AsyncFunction("torShutdown") { notReady<Unit>() }
  }
}
